# qc_wgbs

**WGBS epigenetics QA/QC pipeline**

**Authors:** Derek Ng and Darrell O. Ricke, Ph.D.  (Darrell.Ricke@ll.mit.edu )
  Copyright:  Copyright (c) 2023 Massachusetts Institute of Technology 
  License:    GNU GPL license (http://www.gnu.org/licenses/gpl.html)  

**RAMS request ID 1021178**

**Overview:**
This is the Whole Genome Bisulfite Sequencing (WGBS) epigenetics QA/QC pipeline.

**Citation:** None

**Disclaimer:**
DISTRIBUTION STATEMENT A. Approved for public release. Distribution is unlimited.

This material is based upon work supported by the Defense Advanced Research 
Projects Agency under Air Force Contract No. (FA8702- 15-D-0001). Any opinions, 
findings and conclusions or recommendations expressed in this material are 
those of the author(s) and do not necessarily reflect the views of the 
Defense Advanced Research Projects Agency.

Â© 2023 Massachusetts Institute of Technology.

The software/firmware is provided to you on an As-Is basis

Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
Part 252.227-7013 or 7014 (Feb 2014). Notwithstanding any copyright notice,
U.S. Government rights in this work are defined by DFARS 252.227-7013 or
DFARS 252.227-7014 as detailed above. Use of this work other than as specifically
authorized by the U.S. Government may violate any copyrights that exist in this work.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


## Instructions

1) Please build the [qc_base] container first before proceeding.

2) Build NRSwgbs.def: 

    singularity build NRSwgbs.sif NRSwgbs.def

3) Put your input data files in the "io/" directory.

4) Build wgbs.def: 

    singularity build wgbs.sif wgbs.def

5) Run the container with:
    ```
    singularity run -B io/:/io/ wgbs.sif -1 <read_file1> [options]
    ```
   Output generated by the pipeline will be stored in the "io" directory.

## Run options

	    -1 <string>    The first input read filename (required option; don't type io/<filenam>])

	    -2 <string>    The second input read filename (only if paired-end)

	    -a <string>    Sequence of an adapter ligated to the 3' end of the first read (default = "AGATCGGAAGAGC")

	    -b <string>    3' adapter to be removed from second read in a pair (default = "AGATCGGAAGAGC")

	    -t <int>       The number of threads (default = 16)

	    -h             Print this help message

## Metrics
```
    Genome coverage
    C to T converstion rate (Bisulfite conversion rate)
    %uniquely_mapped
    PCR duplicates rate
    Genome GC content (100 bp bins on chromosome 1)
    Mean trimmed sequence length
```
