# sc_rna_seq


**scRNA-seq transcriptomics QA/QC pipeline on CellRanger results**

**Authors:** Derek Ng and Darrell O. Ricke, Ph.D.  (Darrell.Ricke@ll.mit.edu )
  Copyright:  Copyright (c) 2023 Massachusetts Institute of Technology 
  License:    GNU GPL license (http://www.gnu.org/licenses/gpl.html)  

**RAMS request ID 1021178**

**Overview:**
This is the scRNA-seq transcriptomics QA/QC pipeline on CellRanger results.

**Citation:** None

**Disclaimer:**
DISTRIBUTION STATEMENT A. Approved for public release. Distribution is unlimited.

This material is based upon work supported by the Defense Advanced Research 
Projects Agency under Air Force Contract No. (FA8702- 15-D-0001). Any opinions, 
findings and conclusions or recommendations expressed in this material are 
those of the author(s) and do not necessarily reflect the views of the 
Defense Advanced Research Projects Agency.

Â© 2023 Massachusetts Institute of Technology.

The software/firmware is provided to you on an As-Is basis

Delivered to the U.S. Government with Unlimited Rights, as defined in DFARS
Part 252.227-7013 or 7014 (Feb 2014). Notwithstanding any copyright notice,
U.S. Government rights in this work are defined by DFARS 252.227-7013 or
DFARS 252.227-7014 as detailed above. Use of this work other than as specifically
authorized by the U.S. Government may violate any copyrights that exist in this work.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.


## Instructions

1) Please build the [qc_base] container first before proceeding.

2) Build NRSscRNAseq.def: 

    singularity build NRSscRNAseq.sif NRSscRNAseq.def

3) Put your input data files in the "io/" directory.

4) Build scRNAseq.def: 

    singularity build scRNAseq_ranger.sif scRNAseq_ranger.def

5) Run the container with:
    
    singularity run -B io/:/io/ scRNAseq_ranger.sif -1 <read_file1> [options]
    
   Output generated by the pipeline will be stored in the "io/" directory.

## Run options

	    -1 <string>    The first input read filename (required option; don't type io/<filename>)

	    -2 <string>    The second input read filename (only if paired-end)

	    -t <int>       The number of threads (default = 16)

	    -h             Print this help message

## Metrics scRNA-seq
```
    Estimated number of cells
    Median UMI per cell
    Reads mapped to exons (%exon)
    Fraction reads in cells (frac_cells)
    %cells with mitochondrial reads >= 20%
    Mean trimmed sequence length
```

